### ISSUE TITLE
What is the difference between ds and fsdp?

### ISSUE THREAD

**Issue by RZFan525** [SparkJiao/pytorch-transformers-template#1]
What is the difference between ds and fsdp?

**Comment by SparkJiao** [comment]
I think the implementation of FSDP and Deepspeed is different.

For my own experience, fsdp enables zero-3 techniques but I failed to enable the cpu-offload technique. But the codebase using deepspeed can enable cpu-offload successfully with zero-2. But I failed to use zero-3 with deepspeed.

Generally, I recommend fsdp since its implementation is easy to understand. Deepspeed supports more complex techniques beyond data parallel. FSDP is a tool for data parallel only and I didn't try other techniques of fairscale currently.

**Comment by SparkJiao** [comment]
BTW, deepspeed + zero-2 + cpu-offload may allow you to train large model on single GPU, which cannot be done using FSDP (I just cannot run it successfully by myself).

**Comment by RZFan525** [comment]
Thank you for your reply and detailed answers.

I also would like to ask if you could give me some suggestions when I just want to finetuning PLMs(e.g. BERT\T5\BART) in downstream tasks for research (Maybe using many models and many datasets).  Which one suits me better?

**Comment by SparkJiao** [comment]
If you plan to use this template, I could give a brief introduction about how to use it (I didn't write the necessary readme).

The trainers between deepspeed and fsdp have no difference concerned with the other modules, e.g., the data processing or model. Training loop is decoupled with other modules through hydra, which can configure different components in the config file instead of hard coding for calling interface.

For different downstream tasks, what you should write by yourself are:
1. Data loading. Just ensure that the function specified by the `read_tensor` field in the config file should return a dataset overriding `torch.utils.data.Dataset`, [since which is the input to dataloader](https://github.com/SparkJiao/pytorch-transformers-template/blob/main/trainer_base_fsdp_v1.py#L110). If you need a specific batch collator, just write it by yourself and specify it in the `collator` field.
2. Model. Any huggingface model is OK. You may refer to the interface I used in the examples to learn how to use the metrics.
3. Metrics if you need. Metric class defined as [the class](https://github.com/SparkJiao/pytorch-transformers-template/blob/main/general_util/metrics.py) or simple calculated coupled in the [model implementation](https://github.com/SparkJiao/pytorch-transformers-template/blob/main/models/roberta.py#L21-L116) are both OK. (I have just uploaded the examples of roberta and mt5 for your reference).

**Comment by SparkJiao** [comment]
I also uploaded a example for [data loading](https://github.com/SparkJiao/pytorch-transformers-template/blob/main/data/strategy_qa.py) and corresponding [config](https://github.com/SparkJiao/pytorch-transformers-template/blob/main/conf/roberta_split_fact_v1_1.yaml) for your reference.

**Comment by RZFan525** [comment]
Thank you  !!!

I am going to learn and use your template. This is helpful for me.

Thank you again :) 

Best wishes
